### PR TITLE
HungPV - FinalProject

### PR DESCRIPTION
**Pull request by at-hungpham**
Update UI & Tabbar For Final_Project_YouTube

### CODE REVIEW

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import Foundation
+import RealmSwift
+
+final class RealmManager {
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import Foundation
+import RealmSwift
+
+final class RealmManager {
+    
+    private init() { }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import Foundation
+import RealmSwift
+
+final class RealmManager {
+    
+    private init() { }
+    
+    static var shared = RealmManager()
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            return try Realm(fileURL: url)
+        } catch {
+            return nil
+        }
+    }()
+}
+
+// Mark : - Method
+extension RealmManager {
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        closesure()
+        do {
+            try realm?.commitWrite()
+            completion?(.success)
+        } catch {
+            realm?.cancelWrite()
+            completion?(.failure(error))
+        }
+    }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        }
+    }
+    
+    func fetchObjects<T: Object>(_ type: T.Type, filter predicate: NSPredicate? = nil) -> Results<T>? {
+        guard let predicate = predicate else {
+            return realm?.objects(type)
+        }
+        return realm?.objects(type).filter(predicate)
+    }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            return realm?.objects(type)
+        }
+        return realm?.objects(type).filter(predicate)
+    }
+    
+    func fetchObject<T: Object>(_ type: T.Type, filter predicate: NSPredicate? = nil) -> T? {
+        let results = fetchObjects(type, filter: predicate)
+        return results?.first
+    }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        let results = fetchObjects(type, filter: predicate)
+        return results?.first
+    }
+    
+    func add<T: Object>(object: T, completion: ((RealmResult) -> Void)? = nil) {
+        write({
+            realm?.add(object, update: true)
+        }, completion: completion)
+    }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            realm?.add(object, update: true)
+        }, completion: completion)
+    }
+    
+    func add<T: Object>(objects: [T], completion: ((RealmResult) -> Void)? = nil) {
+        write({
+            realm?.add(objects, update: true)
+        }, completion: completion)
+    }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            realm?.add(objects, update: true)
+        }, completion: completion)
+    }
+    
+    func delete<T: Object>(object: T, completion: ((RealmResult) -> Void)? = nil) {
+        write({
+            realm?.delete(object)
+        }, completion: completion)
+    }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            realm?.delete(object)
+        }, completion: completion)
+    }
+    
+    func deleteAll<T: Object>(objects: [T], completion: ((RealmResult) -> Void)? = nil) {
+        write({
+            realm?.delete(objects)
+        }, completion: completion)
+    }
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/RealmManager/RealmManager.swift [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    func observe<T: Object>(type: T.Type, completion: @escaping (RealmCollectionChange<Results<T>>) -> Void) -> NotificationToken? {
+        let notificationToken = realm?.objects(type).observe({ (change) in
+            completion(change)
+        })
+        return notificationToken
+    }
+}
+
+extension RealmManager {
+
```
Comment (at-bienle) [inline FinalProject/Model/RealmManager/RealmManager.swift]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/AppDelegate/AppDelegate.swift [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+import UIKit
+
+@UIApplicationMain
+class AppDelegate: UIResponder, UIApplicationDelegate {
+
+    var window: UIWindow?
+
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+
```
Comment (at-bienle) [inline FinalProject/AppDelegate/AppDelegate.swift]: WARNING - There should be a space after //

**Comment by at-bienle** [comment]
Total linter violations in pull request: 13
Serious: 0
Warnings: 13

File: FinalProject/Model/CoreApi/Core/ApiManager.swift [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+extension ApiManager.Path {
+    struct Snippet: ApiPath {
+        static var path: String { return baseURL / "v3/videos?" }
+        let chart: String
+        let regionCode: String
+        let maxResults: Int
+        let keyID: String
+        var urlString: String { return Snippet.path + "part=snippet,contentDetails,statistics&chart=\(chart)&regionCode=\(regionCode)&maxResults=\(maxResults)&key=\(keyID)"}
```
Comment (at-bienle) [inline FinalProject/Model/CoreApi/Core/ApiManager.swift]: WARNING - Closure expressions should have a single space inside each brace.

File: FinalProject/Model/CoreApi/API.Downloader.swift [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+//
+
+import Foundation
+import UIKit
+
+extension ApiManager.Downloader {
+    static func downloadImage(imageURL: String, index: Int, completion: @escaping (UIImage?, Int) -> Void) {
+        API.shared().request(urlString: imageURL) { (result) in
+            switch result {
+            case .failure(_):
```
Comment (at-bienle) [inline FinalProject/Model/CoreApi/API.Downloader.swift]: WARNING - Arguments can be omitted when matching enums with associated types if they are not used.

**Comment by at-bienle** [comment]
Total linter violations in pull request: 5
Serious: 0
Warnings: 5

**Comment by at-tienle** [comment]
Không để ý, chừ vào mới thấy 1 cái `pull request` chất lượng ghê. @at-hungpham em tách nhỏ ra nhiều `pull request` chứ em tính làm hết `project` vào 1 pull request à. Lại sử dụng git yếu quá.

**Comment by at-bienle** [comment]
More files for review

**Pull request closed without merging**